src/routes/login/post.rs

```diff
@@ -0,0 +1,94 @@
+use std::sync::Arc;
+
+use axum::{
+    extract::{Form, State},
+    http,
+    response::IntoResponse,
+};
+use hmac::{Hmac, Mac};
+use secrecy::{ExposeSecret, Secret};
+
+use crate::{
+    authentication::{validate_credentials, AuthError, Credentials},
+    startup::AppState,
+    util::error_chain_fmt,
+};
+
+#[derive(serde::Deserialize)]
+pub struct FormData {
+    username: String,
+    password: Secret<String>,
+}
+
+#[derive(thiserror::Error)]
+pub enum LoginError {
+    #[error("Authentication failed")]
+    Auth(#[source] anyhow::Error),
+    #[error("Something went wrong")]
+    Unexpected(#[from] anyhow::Error),
+}
+
+impl std::fmt::Debug for LoginError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        error_chain_fmt(self, f)
+    }
+}
+
+impl IntoResponse for LoginError {
+    fn into_response(self) -> axum::response::Response {
+        match self {
+            LoginError::Unexpected(_) => http::StatusCode::INTERNAL_SERVER_ERROR.into_response(),
+            // ! Not sure we even need this or should be impl IntoResponse
+            LoginError::Auth(_) => http::StatusCode::UNAUTHORIZED.into_response(),
+        }
+    }
+}
```

Comment:
Could prob play around with deleting this when I finish the book. Im not sure I need it anymore since I am returning an `impl IntoResponse` instead of a `LoginError` specifically